presto-main/src/main/java/com/facebook/presto/execution/buffer/BroadcastOutputBuffer.java

```diff
@@ -263,9 +275,17 @@ public void setNoMorePages()
     @Override
     public void destroy()
     {
+        checkState(!Thread.holdsLock(this), "Can not destroy while holding a lock on this");
+
         // ignore destroy if the buffer already in a terminal state.
         if (state.setIf(FINISHED, oldState -> !oldState.isTerminal())) {
-            noMoreBuffers();
+            // dereference initial pages outside of synchronized to avoid making a callback while holding a lock
+            List<SerializedPageReference> pages;
+            synchronized (this) {
+                pages = ImmutableList.copyOf(initialPagesForNewBuffers);
+                initialPagesForNewBuffers.clear();
+            }
+            pages.forEach(SerializedPageReference::dereferencePage);
```

Comment:
Note to self: this effectively inlined `noMoreBuffers` but removed check that all created buffers are declared.